### PR TITLE
Extend SDF conversion

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
Review commits separately

Used to generate https://github.com/tue-robotics/ed_tutorials/pull/10

### CODE REVIEW

**Review by PetervDooren — CHANGES_REQUESTED** [review]

File: src/ed_object_models/conversion_sdf.py [hunk 1]
```diff
@@ -91,6 +118,21 @@ def read_pose(yaml_dict: Mapping) -> str:
     return " ".join(map(str, pose))
 
 
+def create_heightmap(model_name: str, image_path: str, resolution: float, height: float, x: float, y: float) -> str:
+    mesh_path = f"{path.splitext(image_path)[0]}.stl"
+    try:
+        check_call(
+            f"rosrun ed ed_heightmap_to_mesh {image_path} {mesh_path} {resolution} {height} {x} {y}",
+            executable="/bin/bash",
+            shell=True,
```
Comment (PetervDooren) [inline src/ed_object_models/conversion_sdf.py:127]: Is there no better way of doing this? Where do we want to keep all the ed model tooling? in ed or ed_object models?

File: src/ed_object_models/conversion_sdf.py [hunk 2]
```diff
@@ -253,8 +305,15 @@ def read_shape(shape: Mapping, link_names: List[str], color: OrderedDict, model_
     """
     sdf_link = []
     # Check if compound type has name (inserted as comment in yaml)
-    if "compound" in shape:
-        for item in shape["compound"]:
+    keys = [key for key in ["compound", "group"] if key in shape]
```
Comment (PetervDooren) [inline src/ed_object_models/conversion_sdf.py:308]: I don't think its a good idea to support two formats for this function. What new files use 'group' and why cant they use 'compound'

File: src/ed_object_models/conversion_sdf.py [hunk 3]
```diff
@@ -26,6 +28,15 @@ class bcolors:
...
 
+@contextmanager
+def last_model_file(model_file: str):
+    global LAST_MODEL_FILE
+    old_last_model_file = LAST_MODEL_FILE
+    LAST_MODEL_FILE = model_file
+    yield
+    LAST_MODEL_FILE = old_last_model_file
+
+
```
Comment (PetervDooren) [inline src/ed_object_models/conversion_sdf.py:39]: What is the purpose of this?

File: src/ed_object_models/conversion_sdf.py [hunk 4]
```diff
@@ -73,11 +73,11 @@ def unique_name(name: str, names: List[str]) -> str:
     return name
 
 
-def read_pose(yaml_dict: dict) -> str:
+def read_pose(yaml_dict: Mapping) -> str:
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/ed_object_models/conversion_sdf.py:103]: As we need the properties of a Mapping object, not strictly a dict. I will change the variable name.

File: scripts/generate_qr_codes.py [hunk 5]
```diff
@@ -7,14 +7,15 @@
 
 import os
 import rospkg
+
 rospack = rospkg.RosPack()
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply scripts/generate_qr_codes.py:11]: We are indeed not using them right now. But I think this is still useful code.

File: src/ed_object_models/conversion_sdf.py (hunk 1 above)
Comment (MatthijsBurgh) [reply src/ed_object_models/conversion_sdf.py:127]: The `ed_object_models` pkg is only for TUe. This tool is useful for other people too.

File: src/ed_object_models/conversion_sdf.py (hunk 2 above)
Comment (MatthijsBurgh) [reply src/ed_object_models/conversion_sdf.py:308]: This has always been in the cpp code of ed. The convertor was therefore incomplete.

File: src/ed_object_models/conversion_sdf.py (hunk 3 above)
Comment (MatthijsBurgh) [reply src/ed_object_models/conversion_sdf.py:39]: As models can be defined in their own file, but also in a shared file. Therefore a model path doesn't always exist for a model. So when resolving "$(file FILENAME)", you need the path of the file it is in. As the the model path isn't available, we need to track the last opened file.

**Review by PetervDooren — APPROVED** [review]

**Pull request merged**